service/consul/datacenterWatch.go

```diff
@@ -1,109 +1,199 @@
 
+	} else if options.ChrysomConfig != nil && options.ChrysomConfig.PullInterval <= 0 {
+		return nil, errors.New("chrysom pull interval cannot be 0")
+	}
+
+	//start consul watch
+	ticker := time.NewTicker(datacenterWatcher.consulWatchInterval)
+	go datacenterWatcher.watchDatacenters(ticker)
+
+	return datacenterWatcher, nil
+
+}
+
+func (d *datacenterWatcher) stop() {
+	if d.chrysomClient != nil {
+		d.chrysomClient.Stop(context.Background())
+	}
+}
+
+func (d *datacenterWatcher) watchDatacenters(ticker *time.Ticker) {
 	for {
 		select {
-		case <-i.shutdown:
+		case <-d.environment.Closed():
+			ticker.Stop()
+			d.stop()
 			return
-		case <-checkDatacenters.C:
-			currentInstancers := environment.Instancers()
-			keys := make(map[string]bool)
-			instancersToAdd := make(service.Instancers)
-
-			datacenters, err := getDatacenters(logger, client, options)
+		case <-ticker.C:
+			datacenters, err := getDatacenters(d.logger, d.environment.Client(), d.options)
 
 			if err != nil {
 				continue
 			}
 
-			for _, w := range options.watches() {
-				if w.CrossDatacenter {
-					for _, datacenter := range datacenters {
-						w.QueryOptions.Datacenter = datacenter
-
-						// create keys for all datacenters + watched services
-						key := newInstancerKey(w)
-						keys[key] = true
-
-						// don't create new instancer if it is already saved in environment's instancers
-						if currentInstancers.Has(key) {
-							continue
-						}
-
-						// don't create new instancer if it was already created and added to the new instancers map
-						if instancersToAdd.Has(key) {
-							logger.Log(level.Key(), level.WarnValue(), logging.MessageKey(), "skipping duplicate watch", "service", w.Service, "tags", w.Tags, "passingOnly", w.PassingOnly, "datacenter", w.QueryOptions.Datacenter)
-							continue
-						}
-
-						// create new instancer and add it to the map of instancers to add
-						instancersToAdd.Set(key, newInstancer(logger, client, w))
-					}
+			d.updateInstancers(datacenters)
+
+		}
+
+	}
+}
+
+func (d *datacenterWatcher) updateInstancers(datacenters []string) {
+	keys := make(map[string]bool)
+	instancersToAdd := make(service.Instancers)
+
+	currentInstancers := d.environment.Instancers()
+
+	for _, w := range d.options.watches() {
+		if w.CrossDatacenter {
+			for _, datacenter := range datacenters {
+
+				//check if datacenter is part of inactive datacenters list
+				d.lock.RLock()
+				if _, found := d.inactiveDatacenters[datacenter]; found {
+					d.logger.Log(level.Key(), level.InfoValue(), logging.MessageKey(), "datacenter set as inactive", "datacenter name: ", datacenter)
+					continue
+				}
+				d.lock.RUnlock()
+
+				w.QueryOptions.Datacenter = datacenter
+
+				// create keys for all datacenters + watched services
+				key := newInstancerKey(w)
+				keys[key] = true
+
+				// don't create new instancer if it is already saved in environment's instancers
+				if currentInstancers.Has(key) {
+					continue
```

Comment:
All these `continue`s give me the feeling that this code in the `for` loop would be better as its own function.  Not sure how easy that would be with various variables, but it might be worth a try! I think it might make the code easier to understand.